src/app/games/tafl/TaflRules.ts

```diff
@@ -0,0 +1,462 @@
+        return GameStatus.ONGOING;
+    }
+    public getWinner(state: S): MGPOptional<Player> {
+        const LOCAL_VERBOSE: boolean = false;
+        const optionalKingCoord: MGPOptional<Coord> = this.getKingCoord(state.board);
+        if (optionalKingCoord.isAbsent()) {
+            display(LOCAL_VERBOSE, 'The king is dead, victory to invader');
+            // the king is dead, long live the king
+            return MGPOptional.of(this.getInvader());
+        }
+        const kingCoord: Coord = optionalKingCoord.get();
+        if (this.isExternalThrone(kingCoord)) {
+            display(LOCAL_VERBOSE, 'The king escape, victory to defender');
+            // king reached one corner!
+            return MGPOptional.of(this.getDefender());
+        }
+        if (this.isPlayerImmobilized(Player.ZERO, state)) {
+            display(LOCAL_VERBOSE, 'Zero has no move, victory to one');
+            return MGPOptional.of(Player.ONE);
+        }
+        if (this.isPlayerImmobilized(Player.ONE, state)) {
+            display(LOCAL_VERBOSE, 'One has no move, victory to zero');
+            return MGPOptional.of(Player.ZERO);
+        }
+        display(LOCAL_VERBOSE, 'no victory');
+        return MGPOptional.empty();
+    }
+    public getKingCoord(board: Table<TaflPawn>): MGPOptional<Coord> {
+        display(TaflRules.VERBOSE, { TablutRules_getKingCoord: { board, con: this.config } });
+
+        for (let y: number = 0; y < this.config.WIDTH; y++) {
+            for (let x: number = 0; x < this.config.WIDTH; x++) {
+                if (board[y][x].isKing()) {
+                    return MGPOptional.of(new Coord(x, y));
+                }
+            }
+        }
+        return MGPOptional.empty();
+    }
+    public getInvader(): Player {
+        return this.config.INVADER;
+    }
+    public getDefender(): Player {
+        return this.config.INVADER.getOpponent();
+    }
+    public isPlayerImmobilized(player: Player, state: S): boolean {
+        return this.getPlayerListMoves(player, state).length === 0;
+    }
+    public getPlayerListMoves(player: Player, state: S): TaflMove[] {
+        const LOCAL_VERBOSE: boolean = false;
+        const listMoves: TaflMove[] = [];
+        const listPawns: Coord[] = this.getPlayerListPawns(player, state.board);
+        display(TaflRules.VERBOSE || LOCAL_VERBOSE, { TablutRules_getPlayerListMoves: { player, state } });
+
+        let pawnDestinations: Coord[];
+        let newMove: TaflMove;
+        for (const pawn of listPawns) {
+            pawnDestinations = this.getPossibleDestinations(pawn, state);
+            for (const destination of pawnDestinations) {
+                newMove = this.moveGenerator(pawn, destination);
+                listMoves.push(newMove);
+            }
+        }
+        return listMoves;
+    }
+    public getPlayerListPawns(player: Player, board: Table<TaflPawn>): Coord[] {
+        const listPawn: Coord[] = [];
+        let pawn: Coord;
+        let owner: Player;
```

Comment:
const in loop!

Reply:
mAiS c'EsT pLuS oPtImIsÉ pOuR lEs aLlOcAtIoNs De MeMoIrEs !

Reply:
je demande à voir les données qui montrent ça ! (à mon avis le compilo se rend bien compte qu'il est pas nécessaire d'allouer quoi que ce soit, un const est généralement assez bien optimisé) 
(tu peux résoudre!)